### PR TITLE
[cinder-csi-plugin] find device by path from lun address

### LINKED ISSUE #1192: [cinder-csi-plugin] Failed to mount volume to pod on HyperV VM
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
The volume is not mounted to the pod and got error message, `Unable to find Device path for volume`. After changing the log level of the node plugin to `-v=4` I get this in the log:
```
I0908 22:19:22.200863       1 mount.go:178] Failed to find device for the volumeID: "9dc29d58-3d78-47ff-b976-043d7e0be5b5" by serial ID
I0908 22:19:22.311823       1 metadata.go:187] Attempting to fetch metadata from http://169.254.169.254/openstack/latest/meta_data.json
I0908 22:19:23.940241       1 metadata.go:219] Found disk metadata for volumeID "9dc29d58-3d78-47ff-b976-043d7e0be5b5". Bus: "scsi", Address: "0:0:0:2"
I0908 22:19:23.940333       1 metadata.go:236] expecting to find one disk path for volumeID "9dc29d58-3d78-47ff-b976-043d7e0be5b5", found 0: []
I0908 22:19:23.940345       1 metadata.go:242] Could not retrieve device metadata for volumeID: "9dc29d58-3d78-47ff-b976-043d7e0be5b5"
E0908 22:19:23.940353       1 utils.go:84] GRPC error: rpc error: code = Internal desc = Unable to find Device path for volume
```
I can see in Openstack that the volume is created in the correct availability zone and attached to the node. The disk shows up on the node as `dev/sdc` and the PVC get bound to a PV.
When looking in the code at https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/util/metadata/metadata.go#L223
I see that it search for a disk with pattern `/dev/disk/by-path/*-%s-%s", device.Bus, device.Address`
The metadata in openstack gives this info about the disk `{"type": "disk", "serial": "9dc29d58-3d78-47ff-b976-043d7e0be5b5", "bus": "scsi", "address": "0:0:0:2", "tags": []}`
and that will make the pattern compile to `/dev/disk/by-path/*-scsi-0:0:0:2`
when looking on that path on the node I see that `/dev/sdc` has another pattern
`acpi-VMBUS:00-vmbus-8d38ceb678f042ddb67407c98652e9d1-lun-2 -> ../../sdc`

**What you expected to happen**:
Expected behavior is that the volume is mounted to the pod. 

**How to reproduce it**:
Created a new cluster with Magnum. Use the [manifests](https://github.com/kubernetes/cloud-provider-openstack/tree/master/manifests/cinder-csi-plugin) in this repo to deploy the cinder-csi-plugin. I also enabled Topology support (`--feature-gates=Topology=true` as argument to csi-provisioner). Otherwise the volume didn't end up in the correct availabilty zone and it failed. Create a storage class.
Create a PVC using that storageclass and create a pod that mounts that PVC.
See below for the environment to use.

**Environment**:
- openstack-cloud-controller-manager(or other related binary) version:
- OpenStack version: Stein
- Openstack hypervisor: Hyper-V
- OS image: fedora-coreos 32.20200809.3.0
- Magnum: Train
- Kubernetes: 1.18.2
- Cloud-provider: 1.18.0
- Images (via the manifests):
  - docker.io/k8scloudprovider/cinder-csi-plugin:latest
  - quay.io/k8scsi/csi-provisioner:v1.6.0

### PR DESCRIPTION
**Pull request by peterrosell**
Disks can appear with the pattern `/dev/disk/by-path/*lun-<id>` on Hyper-V.
Using the last value in the address as an id.

**What this PR does / why we need it**:
Makes it possible to detect disks on Hyper-V that appear with the pattern  `/dev/disk/by-path/*lun-<id>`.

**Which issue this PR fixes(if applicable)**:
fixes #1192

**Special notes for reviewers**:
To test this PR you need Openstack installed on top of Hyper-V and install kube and cinder-csi-plugin. Then create a matching storage class, a PVC using that storage class and a Pod that mounts the PVC. See the issue for more details.

**Release note**:

```release-note
[cinder-csi-plugin] Add support for detecting disks by-path for lun volumes on Hyper-V.
```

### CODE REVIEW

**Comment by jichenjc** [comment]
@ramineni please help to comment

File: pkg/util/metadata/metadata.go [hunk 1]
```diff
@@ -233,6 +233,23 @@ func GetDevicePath(volumeID string) string {
 				return diskPaths[0]
 			}
 
+			adressParts := strings.Split(device.Address, ":")
```
Comment (jichenjc) [inline pkg/util/metadata/metadata.go:236]: can we add some comment here so that we can understand why we made changes here?

Comment (peterrosell) [reply pkg/util/metadata/metadata.go:239]: I added a comment with ref to the udev code that creates the soft link.

**Comment by jichenjc** [comment]
/lgtm

thanks~

**Comment by peterrosell** [comment]
I have been playing around more with this and found a case where several disk have the same lun number. I don't know why this happened. Will try to get some feedback from our hosting company and see if they did some changes.
It looks like this
```
capi-quickstart-md-0-6zcsk:/dev/disk/by-path$ ls -l
total 0
lrwxrwxrwx 1 root root  9 Sep 17 23:19 acpi-VMBUS:01-vmbus-00000000000088990000000000000000-lun-0 -> ../../sdb
lrwxrwxrwx 1 root root 10 Sep 17 23:19 acpi-VMBUS:01-vmbus-00000000000088990000000000000000-lun-0-part1 -> ../../sdb1
lrwxrwxrwx 1 root root 11 Sep 17 23:19 acpi-VMBUS:01-vmbus-00000000000088990000000000000000-lun-0-part14 -> ../../sdb14
lrwxrwxrwx 1 root root 11 Sep 17 23:19 acpi-VMBUS:01-vmbus-00000000000088990000000000000000-lun-0-part15 -> ../../sdb15
lrwxrwxrwx 1 root root  9 Sep 17 23:19 acpi-VMBUS:01-vmbus-00000001000088990000000000000000-lun-0 -> ../../sda
lrwxrwxrwx 1 root root  9 Sep 17 23:19 acpi-VMBUS:01-vmbus-6a26378e3d28433c8593c754e39ed66b-lun-0 -> ../../sdc
```
i don't know where the guid is created and if it's possble to get from the meta data. I will try to do some more investigations.

**Comment by jichenjc** [comment]
is it becaues of multipath setup ? no idea on hyper-v implementation...

**Comment by peterrosell** [comment]
I don't know about multipath. Before the disk showed up here
```
/sys/devices/LNXSYSTM:00/LNXSYBUS:00/ACPI0004:00/VMBUS:00/216abb18-f151-45cc-a3b5-a981b815575a/host0/target0:0:0/0:0:0:2
```
This time it showed up at
```
./LNXSYBUS:00/PNP0A03:00/device:07/VMBUS:01/6a26378e-3d28-433c-8593-c754e39ed66b/host4/target4:0:0/4:0:0:0
```
I noticed that when I add more disks they get the same guid (6a....66b) and the last number is set to 1 which matches the lun-number that shows up under `by-path`. 
So maybe the disk with a lot of zeros are "special" cases. I will try to find some info about this, but at the moment I'm a bit busy with other stuff.

**Comment by capsali** [comment]
I think a more accurate disk pinpoint would be from "/sys/bus/scsi/devices/<scsi_address_from_metadata>/block". 
This is an issue on CoreOS, since Ubuntu has the "/dev/disk/by-path" include the scsi address in full.
For the issue with different guids and same LUNs, is the VM gen1 or gen2? This looks like being a gen1 VM.

**Comment by jichenjc** [comment]
I doubt whether we tested CoreOS though I knew openshift is intend to support OCCM
https://github.com/openshift/enhancements/pull/423 soon, so maybe we need focus on Hyper-V for this issue and open independent issue for CoreOS

**Comment by ramineni** [comment]
> I have been playing around more with this and found a case where several disk have the same lun number. I don't know why this happened. Will try to get some feedback from our hosting company and see if they did some changes.
> It looks like this
> 
> ```
> capi-quickstart-md-0-6zcsk:/dev/disk/by-path$ ls -l
> total 0
> lrwxrwxrwx 1 root root  9 Sep 17 23:19 acpi-VMBUS:01-vmbus-00000000000088990000000000000000-lun-0 -> ../../sdb
> lrwxrwxrwx 1 root root 10 Sep 17 23:19 acpi-VMBUS:01-vmbus-00000000000088990000000000000000-lun-0-part1 -> ../../sdb1
> lrwxrwxrwx 1 root root 11 Sep 17 23:19 acpi-VMBUS:01-vmbus-00000000000088990000000000000000-lun-0-part14 -> ../../sdb14
> lrwxrwxrwx 1 root root 11 Sep 17 23:19 acpi-VMBUS:01-vmbus-00000000000088990000000000000000-lun-0-part15 -> ../../sdb15
> lrwxrwxrwx 1 root root  9 Sep 17 23:19 acpi-VMBUS:01-vmbus-00000001000088990000000000000000-lun-0 -> ../../sda
> lrwxrwxrwx 1 root root  9 Sep 17 23:19 acpi-VMBUS:01-vmbus-6a26378e3d28433c8593c754e39ed66b-lun-0 -> ../../sdc
> ```
> 
> i don't know where the guid is created and if it's possble to get from the meta data. I will try to do some more investigations.

As we dont have a unique way of indentifying the disk as per above comment , keeping this on hold until further investigations

/hold

**Comment by peterrosell** [comment]
@capsali 
I saw that the VM that we run is `hyperv-gen1`, Ubuntu. I tested another image (fedora-coreos) that has `hyperv-gen2` and there the naming of the disk under `by-path` matches how I implemented this PR.
It would be nice to support both gen1 and gen2 images.

**Comment by peterrosell** [comment]
I have added a solution for Hyper-V generation 1 VMs that ignores the root disk and the metadata disk when listing disk under `by-path`. I have assumed that the disks always have the same uuids `00000000000088990000000000000000` and `00000001000088990000000000000000`. I think we can go with that and if someone finds another solution for this or have better knowledge of how it works we can patch it then.

**Comment by jichenjc** [comment]
> I have added a solution for Hyper-V generation 1 VMs that ignores the root disk and the metadata disk when listing disk under `by-path`. I have assumed that the disks always have the same uuids `00000000000088990000000000000000` and `00000001000088990000000000000000`. I think we can go with that and if someone finds another solution for this or have better knowledge of how it works we can patch it then.

I don't have strong skill on those 0000 xxxxx numbers which looks weird
as this can solve hyper-V gen-1 issue 

maybe better to have a hyper-V CI but that's a much bigger story

/lgtm

**Comment by peterrosell** [comment]
@capsali Do you have any input on this solution? I have a feeling that you know more details about Hyper-V?

**Comment by peterrosell** [comment]
@ramineni I have tested this against the OpenStack environment that I have access to and it works for both gen1 and gen2 VMs so it should be fine to merge it.

**Comment by peterrosell** [comment]
It's okay from me if you want to close this PR as our hosting provider has changed their underlying infrastructure from Hyper-V to KVM so I will not put anymore work on it.

**Comment by lingxiankong** [comment]
> It's okay from me if you want to close this PR as our hosting provider has changed their underlying infrastructure from Hyper-V to KVM so I will not put anymore work on it.

Thanks for letting us know @peterrosell, I will close this PR for now as there was not much interest in Hyper-V Kubernetes cluster environment. I'm happy to reopen this if someone shows interest in the future.

/close

**Pull request closed without merging**